manifests/register.pp

```diff
@@ -0,0 +1,25 @@
+# @summary Register the host as Foreman
+# @api private
+class foreman::register (
+  Stdlib::Fqdn $foreman_host_name = $facts['networking']['fqdn'],
+) {
+  foreman_host { $foreman_host_name:
+    ensure          => present,
+    base_url        => $foreman::foreman_url,
+    consumer_key    => $foreman::oauth_consumer_key,
+    consumer_secret => $foreman::oauth_consumer_secret,
+    effective_user  => $foreman::oauth_effective_user,
+    ssl_ca          => $foreman::server_ssl_ca,
+    facts           => $facts,
+  }
+
+  foreman_instance_host { $foreman_host_name:
+    ensure          => present,
+    base_url        => $foreman::foreman_url,
+    consumer_key    => $foreman::oauth_consumer_key,
+    consumer_secret => $foreman::oauth_consumer_secret,
+    effective_user  => $foreman::oauth_effective_user,
+    ssl_ca          => $foreman::server_ssl_ca,
+    require         => Foreman_host[$foreman_host_name],
```

Comment:
In the type you can add:
```ruby
autorequire(:foreman_host) do
  [self[:name]]
end
````

Then you can remove this line

Reply:
https://github.com/theforeman/puppet-foreman_proxy/pull/687 tells me that the smart proxy host can also be an autorequire.